File: threadring/threadring.scala
/* The Computer Language Benchmarks Game
   http://benchmarksgame.alioth.debian.org/

   Contributed by Vincent Kraeutler
   updated for 2.8 by Rex Kerr
*/
import scala.actors.Actor
import scala.actors.Actor._

object threadring {

  class Thread(val label: Int) extends Actor {
    var next: Thread = null
    def act() { loop { react {
      case 0 => println(label); System.exit(0)
      case n: Int => next ! n - 1
    }}}
  }

  // create the threads
  val ring = Array.tabulate(503)(i => new Thread(i + 1))

  // hook them up
  ring.foreach(t => {
    t.next = ring( t.label % ring.length )
    t.start
  })

  def main(args : Array[String]) {
    val nHops = args(0).toInt
    ring(0) ! nHops
  }

}

File: threadring/threadring.rb
# The Computer Language Benchmarks Game
# http://benchmarksgame.alioth.debian.org
# contributed by Klaus Friedel
#

#require "thread"

THREAD_COUNT = 503

class Receiver
  @next
  @mailbox

  def initialize(name)
    @name = name
    @mailbox = Queue.new
  end

  def next=(n)
    @next = n
  end

  def put(msg)
    @mailbox.push(msg)
  end

  def messageloop
    while true
      hopsRemaining = @mailbox.pop
      if(hopsRemaining == 0)
        print @name, "\n"
        exit(0)
      end
      @next.put(hopsRemaining - 1)
    end
  end
end

##########
#  Main  #
##########
receivers = []
for i in 0..THREAD_COUNT-1
  receivers[i] = Receiver.new(i+1)
  if(i > 0)
    receivers[i-1].next = receivers[i]
  end
end
#close the ring
receivers[THREAD_COUNT-1].next = receivers[0]

# start the threads
for i in 0..THREAD_COUNT-1
  Thread.start(i){|k| receivers[k].messageloop}
end

receivers[0].put(ARGV[0].to_i)

sleep

